### PR TITLE
Add debounce to query update

### PR DESCRIPTION
**Pull request by TalMalka123**
Fix the issue with the toggle
Add debounce to onChange func
Fix the issue with infinite warnings on the Mui autocomplete- "value provided to Autocomplete is invalid"

### CODE REVIEW

File: src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx [hunk 1]
```diff
@@ -170,48 +167,39 @@ export const QueryBuilderProvider = ({
...
     setIsDisplayableInSimpleMode(checkIfConditionIsDisplayableInSimpleMode(rootCondition));
   }, [rootCondition]);
 
-  useEffect(() => {
-    setIsSimpleMode(checkIfSimpleMode(queryInput.query));
-  }, [forceCompoundMode, queryInput.query]);
-
   function onToggleQueryMode() {
-    setIsSimpleMode(!isSimpleMode);
+    setIsCompoundModeForced(!isCompoundModeForced);
```
Comment (simonlsk) [inline src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:194]:
```suggestion
    setIsSimpleMode((isSimple) => { 
    if (isSimple) {
    return false
    } else {
    return isDisplayableInSimpleMode
    })
```

**Review by simonlsk — CHANGES_REQUESTED** [review]
Do another round of simplification if possible.

File: src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx [hunk 2]
```diff
@@ -227,14 +221,24 @@ export const QueryBuilderProvider = ({
     return condition;
   };
 
+  const debouncedOnChange = useCallback(
```
Comment (simonlsk) [inline src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:224]: Wouldn't be nicer to just debounce `onChange` in the `ConditionTextField` ? That way only the text keyboard inputs are debounced.

**Review by simonlsk — CHANGES_REQUESTED** [review]
LGTM in general. 
I was thinking maybe to delegate the debounce to the text input component, WDYT?

File: src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx [hunk 3]
```diff
@@ -170,48 +167,45 @@ export const QueryBuilderProvider = ({
...
     setIsDisplayableInSimpleMode(checkIfConditionIsDisplayableInSimpleMode(rootCondition));
   }, [rootCondition]);
 
-  useEffect(() => {
-    setIsSimpleMode(checkIfSimpleMode(queryInput.query));
-  }, [forceCompoundMode, queryInput.query]);
-
   function onToggleQueryMode() {
-    setIsSimpleMode(!isSimpleMode);
+    setIsSimpleMode((isSimpleMode) => {
```
[the remark this answers is not in the document]
Comment (TalMalka123) [reply src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:191]: @simonlsk told me that when I want to change the state, I should not use the variable from the state definition, but use the prev from the func, so there won't be race conditioning @elad-n

[the remark this answers is not in the document]
Comment (simonlsk) [reply src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:191]: https://stackoverflow.com/a/61313526

File: src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx [hunk 4]
```diff
@@ -170,48 +167,45 @@ export const QueryBuilderProvider = ({
...
-    checkIfSimpleMode(queryInput.query)
-  );
-  const [metadataFieldsList, setMetadataFieldsList] =
-    useState<MetadataFieldProps[]>(metadataFields);
-  const [isDisplayableInSimpleMode, setIsDisplayableInSimpleMode] = useState<boolean>(
+  const [rootCondition, setRootCondition] = useState<AndOrMetadataInput>(getInitialQuery);
+  const [isDisplayableInSimpleMode, setIsDisplayableInSimpleMode] = useState<boolean>(() =>
     checkIfConditionIsDisplayableInSimpleMode(queryInput.query)
   );
+  const [isSimpleMode, setIsSimpleMode] = useState<boolean>(isDisplayableInSimpleMode);
```
[the remark this answers is not in the document]
Comment (TalMalka123) [reply src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:174]: I'm not sure I understand. My use-state hooks are at the top. Before them there are just functions that the state uses, so they must be defined before the state itself, for them to be recognized. 
According to the article, it doesn't seem to break any rule cause there is no prior return statement within the component itself (unless I miss something), and the hooks are not within any function/condition.

@elad-n Let me know if I miss something, and what I need to do

File: src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx [hunk 5]
```diff
@@ -170,48 +167,45 @@ export const QueryBuilderProvider = ({
...
+      } else {
+        return isDisplayableInSimpleMode;
+      }
+    });
   }
 
   useEffect(() => {
-    setMetadataFieldsList(metadataFields);
-  }, [metadataFields]);
+    setIsSimpleMode((isSimpleMode) => {
```
[the remark this answers is not in the document]
Comment (TalMalka123) [reply src/components/dagshub/data-engine/queryBuilder/QueryBuilderContext.tsx:201]: I guess its the same comment like before: https://github.com/DagsHub/design-system/pull/88#discussion_r1469474437

File: package.json [hunk 6]
```diff
@@ -4,7 +4,7 @@
     "type": "git",
     "url": "git+https://github.com/dagshub/design-system.git"
   },
-  "version": "0.1.76",
+  "version": "0.1.77",
```
Comment (simonlsk) [inline package.json]: bump

**Review by simonlsk — APPROVED** [review]

**Pull request merged**